pallets/clusters/src/lib.rs

```diff
@@ -0,0 +1,520 @@
+			ensure!(!settings.data.len().is_zero(), Error::<T>::InvalidInput);
+
+			let role_hash = blake2_128(&[&cluster_id.as_slice(), role_name.as_slice()].concat());
+			let cluster = <Clusters<T>>::get(&cluster_id).ok_or(Error::<T>::ClusterNotFound)?;
+
+			// Check that the caller is the owner of the cluster
+			ensure!(
+				who == cluster.owner,
+				Error::<T>::NoPermission
+			);
+
+			// At creation there are no Members and no Rules assigned to a Role
+			let new_role = Role { name: role_name.into_inner(), members: vec![], rules: None };
+
+			// Check that the role does not exists already in the cluster
+			if cluster
+				.roles
+				.iter()
+				.any(|role| new_role.name.eq(&role.name))
+			{
+				return Err(Error::<T>::RoleExists.into())
+			}
+
+			// write
+			<Clusters<T>>::mutate(&cluster_id, |cluster| {
+				let cluster = cluster
+					.as_mut()
+					.expect("Should find the cluster");
+				cluster.roles.push(new_role);
+			});
+
+			<RoleToSettings<T>>::append(role_hash, settings);
+
+			Self::deposit_event(Event::RoleCreated { role_hash });
+			log::trace!("Role {:?} created and associated to cluster {:?}", role_hash, cluster_id);
+
+			Ok(())
+		}
+
+		/// Edit a **Role**.
+		#[pallet::weight(10_000 + T::DbWeight::get().writes(1).ref_time())]
+		pub fn edit_role(
+			origin: OriginFor<T>,
+			role_name: BoundedVec<u8, T::NameLimit>,
+			cluster_id: Hash128,
+			new_members_list: BoundedVec<T::AccountId, T::MembersLimit>,
+			new_settings: RoleSetting,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			ensure!(!role_name.len().is_zero(), Error::<T>::InvalidInput);
+
+			if new_members_list.len().is_zero() &&
+				(new_settings.data.is_empty() || new_settings.name.is_empty())
+			{
+				return Err(Error::<T>::InvalidInput.into())
+			}
+
+			let cluster = <Clusters<T>>::get(&cluster_id).ok_or(Error::<T>::SystematicFailure)?;
+
+			// Check that the caller is the owner of the cluster;
+			ensure!(
+				who == cluster.owner,
+				Error::<T>::NoPermission
+			);
+
+			// Check that the role exists in the cluster and in storage
+			if !cluster
+				.roles
+				.iter()
+				.any(|role| role_name.eq(&role.name))
+			{
+				return Err(Error::<T>::RoleNotFound.into())
+			}
+
+			let role_hash = blake2_128(&[&cluster_id[..], &role_name.as_slice()].concat());
+			let role_name_vec = role_name.into_inner();
+
+			// write
+			<Clusters<T>>::mutate(&cluster_id, |cluster| {
+				let cluster = cluster
+					.as_mut()
+					.expect("Should find the cluster");
+				cluster.roles.iter().position(|x| x.name == role_name_vec).map(|index| {
+					let role = cluster
+						.roles
+						.get(index)
+						.expect("Should find the cluster");
+					let mut members = role.clone().members;
+					for member in new_members_list {
+						members.push(member);
+					}
+				});
+			});
+
+			<RoleToSettings<T>>::mutate(&role_hash, |role| {
+				let role_settings = role
+					.as_mut()
+					.expect("Should find the role settings");
```

Comment:
This could potentially fail, I don't see you fetching role_hash before in a safe manner

Reply:
In fact what if does not exist?

Reply:
`role_hash` is generated from `cluster_id` and `role_name`. Both of them are checked before.
Actually, cluster_id implicitly, if a wrong one is passed and its related cluster is not found, it fails
```rust
get(&cluster_id).ok_or(Error::<T>::SystematicFailure)?;
```
and `role_name`
```rust
ensure!(!role_name.len().is_zero(), Error::<T>::InvalidInput);
```
So, if these two checks pass, a hash is surely generated and the mutate will not fail because of it.

Reply:
The second question.. you are right. I need to add a check for the existence of the hash in the storage that I want to mutate.
I will add it.